src/Features/VisualBasic/Portable/ConvertForEachToFor/VisualBasicConvertForEachToForCodeRefactoringProvider.vb

```diff
@@ -0,0 +1,185 @@
+                Return Nothing
+            End If
+
+            Return (foreachBlock.Statements(0), foreachBlock.Statements(foreachBlock.Statements.Count - 1))
+        End Function
+
+        Protected Overrides Sub ConvertToForStatement(model As SemanticModel, foreachInfo As ForEachInfo, editor As SyntaxEditor, cancellationToken As CancellationToken)
+            cancellationToken.ThrowIfCancellationRequested()
+
+            Dim generator = editor.Generator
+            Dim forEachBlock = foreachInfo.ForEachStatement
+
+            ' trailing triva of expression will be attached to for statement below
+            Dim foreachCollectionExpression = forEachBlock.ForEachStatement.Expression
+            Dim collectionVariable = GetCollectionVariableName(
+                model, generator, foreachInfo, foreachCollectionExpression, cancellationToken)
+
+            ' make sure we get rid of all comments from expression since that will be re-attached to for statement
+            Dim expression = foreachCollectionExpression.WithTrailingTrivia(
+                    foreachCollectionExpression.GetTrailingTrivia().Where(Function(t) t.IsWhitespaceOrEndOfLine()))
+
+            ' and remove all trailing trivia if it is used for cast
+            If foreachInfo.RequireExplicitCastInterface Then
+                expression = expression.WithoutTrailingTrivia()
+            End If
+
+            ' first, see whether we need to introduce New statement to capture collection
+            IntroduceCollectionStatement(model, foreachInfo, editor, type:=Nothing, expression, collectionVariable)
+
+            ' create New index varialbe name
+            Dim indexVariable = If(
+                forEachBlock.Statements.Count = 0,
+                generator.Identifier("i"),
+                CreateUniqueName(foreachInfo.SemanticFacts, model, forEachBlock.Statements(0), "i", cancellationToken))
+
+            ' put variable statement in body
+            Dim bodyStatement = GetForLoopBody(generator, foreachInfo, collectionVariable, indexVariable)
+
+            Dim nextStatement = forEachBlock.NextStatement
+
+            If nextStatement.ControlVariables.Count > 0 Then
+                Contract.Requires(nextStatement.ControlVariables.Count = 1)
+
+                Dim controlVariable As SyntaxNode = nextStatement.ControlVariables(0)
+                controlVariable = generator.IdentifierName(
+                    indexVariable _
+                        .WithLeadingTrivia(controlVariable.GetFirstToken().LeadingTrivia) _
+                        .WithTrailingTrivia(controlVariable.GetLastToken().TrailingTrivia))
+
+                nextStatement = nextStatement.WithControlVariables(
+                    SyntaxFactory.SingletonSeparatedList(controlVariable))
+            End If
+
+            ' create for statement from foreach statement
+            Dim forBlock = SyntaxFactory.ForBlock(
+                SyntaxFactory.ForStatement(
```

Comment:
it seems weird to me that we're not using the original 'for' keyword.

Reply:
also tracking here - https://github.com/dotnet/roslyn/issues/25667

that change require new testing over trivia. so I will leave it for next iteration.